core/drivers/imx_i2c.c

```diff
@@ -0,0 +1,397 @@
+	uint32_t sr = 0;
+
+	while (!timeout_elapsed(tref)) {
+		sr = i2c_io_read8(bid, I2SR);
+		if (sr & I2SR_IAL) {
+			EMSG("bus arbitration lost");
+			i2c_io_write8(bid, I2SR, sr & ~I2SR_IAL);
+			return TEE_ERROR_COMMUNICATION;
+		}
+		if ((*match)(sr)) {
+			if (status)
+				*status = sr;
+			return TEE_SUCCESS;
+		}
+	}
+
+	return TEE_ERROR_BUSY;
+}
+
+static TEE_Result i2c_idle_bus(uint8_t bid)
+{
+	uint8_t tmp = i2c_io_read8(bid, I2CR) & ~I2CR_MSTA;
+	TEE_Result ret = TEE_SUCCESS;
+
+	i2c_io_write8(bid, I2CR, tmp);
+	ret = i2c_sync_bus(bid, &bus_is_idle, NULL);
+	i2c_io_write8(bid, I2SR, 0);
+
+	return ret;
+}
+
+static TEE_Result i2c_write_byte(uint8_t bid, uint8_t byte)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t status = 0;
+
+	i2c_io_write8(bid, I2DR, byte);
+	ret = i2c_sync_bus(bid, &isr_active, &status);
+	i2c_io_write8(bid, I2SR, 0);
+
+	if (ret == TEE_SUCCESS && (status & I2SR_RX_NO_AK))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result i2c_read_byte(uint8_t bid, uint8_t *p)
+{
+	TEE_Result ret = TEE_SUCCESS;
+
+	*p = i2c_io_read8(bid, I2DR);
+	ret = i2c_sync_bus(bid, &isr_active, NULL);
+	i2c_io_write8(bid, I2SR, 0);
+
+	return ret;
+}
+
+static TEE_Result i2c_write_data(uint8_t bid, const uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t tmp = 0;
+
+	if (!len)
+		return TEE_SUCCESS;
+
+	tmp = i2c_io_read8(bid, I2CR) | I2CR_MTX | I2CR_TX_NO_AK;
+	i2c_io_write8(bid, I2CR, tmp);
+
+	while (len--) {
+		ret = i2c_write_byte(bid, *buf++);
+		if (ret != TEE_SUCCESS)
+			return ret;
+	}
+
+	return ret;
+}
+
+static TEE_Result i2c_read_data(uint8_t bid, uint8_t *buf, int len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	uint8_t dummy = 0;
+	uint32_t tmp = 0;
+
+	if (!len)
+		return TEE_SUCCESS;
+
+	tmp = i2c_io_read8(bid, I2CR) & ~I2CR_MTX;
+	tmp = (len == 1) ? tmp | I2CR_TX_NO_AK : tmp & ~I2CR_TX_NO_AK;
+	i2c_io_write8(bid, I2CR, tmp);
+	i2c_io_read8(bid, I2DR);
+
+	ret = i2c_read_byte(bid, &dummy);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	/*
+	 * i.MX 8M Mini APRM Rev2 08/2019:
+	 * a data transfer ends when the master signals a stop; for a master
+	 * receiver to terminate a transfer it must inform the slave transmiter
+	 * by not acknowledging the last data byte. This is done by setting the
+	 * transmit acknowledge bit before reading the next-to-last byte.
+	 */
+	do {
+		if (len == 2) {
+			tmp = i2c_io_read8(bid, I2CR) | I2CR_TX_NO_AK;
+			i2c_io_write8(bid, I2CR, tmp);
+		}
+
+		ret = i2c_read_byte(bid, buf++);
+		if (ret != TEE_SUCCESS)
+			return ret;
+	} while (len--);
+
+	return ret;
+}
+
+/* i.MX 8M Mini APRM Rev2 08/2019: steps previous to a read/write operation  */
+static TEE_Result i2c_init_transfer(uint8_t bid, uint8_t chip)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	uint32_t tmp = 0;
+
+	ret = i2c_idle_bus(bid);
+	if (ret != TEE_SUCCESS)
+		return ret;
+
+	/* enable the interface */
+	i2c_io_write8(bid, I2CR, I2CR_IEN);
+
+	/* 16.1.4.1 start master transaction */
```

Comment:
Could you remove the ```16.1.4.1``` ? This chapter is only valid for imx8m reference manual, but the code is not imx8m specific .
Same for line 285,355,361